hollow/src/main/java/com/netflix/hollow/core/index/HollowPrefixIndex.java

```diff
@@ -0,0 +1,438 @@
+        // note reuse only happens once swap is called and bits are reset
+        if (prefixIndex != null) prefixIndex.recycleMemory(memoryRecycle);
+
+        Tst tst = new Tst(maxOrdinalOfType, cardinalityOfType, cardinalityOfKeyField, memoryRecycle);
+        BitSet ordinals = readStateEngine.getTypeState(type).getPopulatedOrdinals();
+        int ordinal = ordinals.nextSetBit(0);
+        while (ordinal != -1) {
+            String[] keys = getKey(ordinal);
+            for (String key : keys)
+                tst.insert(key, ordinal);
+            ordinal = ordinals.nextSetBit(ordinal + 1);
+        }
+
+        prefixIndex = tst;
+        prefixIndexVolatile = tst;
+        // safe to return previous long arrays on next request for long array.
+        memoryRecycle.swap();
+    }
+
+    /**
+     * Return the key to index in prefix index. Override this method to support tokens for the key. By default keys are indexed as lower case characters.
+     * <pre><code>
+     *     String[] keys = super.getKey(ordinal);
+     *     String[] tokens = keys[0].split(" ")
+     *     return tokens;
+     * </pre></code>
+     *
+     * @param ordinal
+     * @return keys to index.
+     */
+    protected String[] getKey(int ordinal) {
+
+        HollowObjectTypeReadState readState = (HollowObjectTypeReadState) readStateEngine.getTypeState(type);
+        int refOrdinal = ordinal;
+        int position = 0;
+        String tempType = type;
+
+        for (int i = 0; i < fields.length; i++) {
+            position = fieldPositions[i];
+            readState = (HollowObjectTypeReadState) readStateEngine.getTypeState(tempType);
+
+            if (fieldTypes[i].equals(HollowObjectSchema.FieldType.REFERENCE)) {
+                tempType = readState.getSchema().getReferencedType(fields[i]);
+                refOrdinal = readState.readOrdinal(refOrdinal, position);
+            }
+        }
+
+        String key = readState.readString(refOrdinal, position).toLowerCase();
+        return new String[]{key};
+    }
+
+    /**
+     * Query the index to find all the ordinals that match the given prefix.
+     *
+     * @param prefix
+     * @return ordinals of the parent object.
+     */
+    public Set<Integer> query(String prefix) {
+        Tst current = this.prefixIndex;
+        Set<Integer> ordinals;
+        do {
+            ordinals = current.query(prefix);
+        } while (current != this.prefixIndexVolatile);
+        return ordinals;
+    }
+
+    /**
+     * Use this method to keep the index updated with delta changes on the read state engine.
+     * Remember to call stopDeltaUpdates to stop the delta changes.
+     * NOTE: Each delta updates creates a new prefix index and swaps the new with current.
+     */
+    public void listenForDeltaUpdates() {
+        readStateEngine.getTypeState(type).addListener(this);
+    }
+
+    /**
+     * Stop delta updates for this index.
+     */
+    public void stopDeltaUpdates() {
```

Comment:
For consistency with the `HollowPrimaryKeyIndex`, can we use the method name `detachFromDeltaUpdates()`?